### PR TITLE
Graphql

### PR DESCRIPTION
**Pull request by alon6699**

### CODE REVIEW

File: src/cart/cart.ts [hunk 1]
```diff
@@ -1,9 +1,15 @@
 import { Cart } from "./types/cart.type";
+import { Product } from "../models/product.model";
+import { findProductQuery } from "../database/product.queries";
+import { logger } from "../logger/logger";
 
 const carts: Record<string, Cart> = {};
 
-export const addCart = (id: string) => {
-    carts[id] = {};
+export const registerNewCart = (id: string) => {
```
Comment (tomporat247) [inline src/cart/cart.ts:8]: revert - I'll explain why this happens in index.ts

File: src/graphql/index.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+const baseSchema: DocumentNode = gql`
+  type Query
+  type Mutation
+  type Subscription
+`
+
+export const apolloServerConfig: Config = {
+    typeDefs: [baseSchema, productsSchema, ProductTypeDef],
+    resolvers: productsResolvers,
+    context: (context) => {
```
Comment (tomporat247) [inline src/graphql/index.ts:17]: this is called every time and you go around it by checking if it exists inside registerNewCart.

for simplicity you may assume that every user will subscribe to productsUpdate, and you can register the cart when the subscription is made

File: src/graphql/product/product.resolver.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+const productUpdatesSideEffect = (product: Product, deleted: boolean = false) => {
+    product.deleted = deleted;
+    pubSub.publish(PRODUCT_UPDATES, { productUpdates: product });
+    return product;
+};
+
+const updateProductInCartHelper = (product: Product) =>
+    product.limit !== undefined ? productUpdatesSideEffect(product) : product;
+
+const checkoutHandler = (userId: string) =>
```
Comment (tomporat247) [inline src/graphql/product/product.resolver.ts:17]: add return type

File: src/graphql/product/product.resolver.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    checkoutQuery(getCart(userId))
+        .then(() => {
+            resetCart(userId);
+            return true;
+        })
+
+export const productsResolvers = {
+    Query: {
+        getProducts: () => getProductsQuery(),
```
Comment (tomporat247) [inline src/graphql/product/product.resolver.ts:26]: just getProductsQuery

File: src/graphql/product/product.resolver.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        getProduct: (_, { id }) => findProductQuery(id)
+    },
+    Mutation: {
+        createProduct: (_, { product }) => createProductQuery(product).then(productUpdatesSideEffect),
+        updateProduct: (_, { id, product }) => updateProductQuery(id, product).then(productUpdatesSideEffect),
+        deleteProduct: (_, { id }) => deleteProductQuery(id).then(product => productUpdatesSideEffect(product, true)),
+        updateProductInCart: async (_, { id, amount }, { userId }) =>
```
Comment (tomporat247) [inline src/graphql/product/product.resolver.ts:33]: no need for async in both

File: src/graphql/product/product.schema.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { gql } from "apollo-server";
+import { DocumentNode } from "graphql";
+
+export const productsSchema: DocumentNode = gql`
```
Comment (tomporat247) [inline src/graphql/product/product.schema.ts:4]: add `!` to return types

File: src/graphql/product/product.types.schema.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { gql } from 'apollo-server';
+import { DocumentNode } from 'graphql';
+
+export const ProductTypeDef: DocumentNode = gql`
+    type Product {
```
Comment (tomporat247) [inline src/graphql/product/product.types.schema.ts:5]: add ! to all but limit

File: src/cart/cart.ts [hunk 8]
```diff
@@ -28,3 +34,19 @@ export const calculateProductLimit = (productId: string, currentLimit: number):
     Object.values(carts).reduce((acc: number, cart: Cart) =>
         cart[productId] ? acc - cart[productId] : acc, currentLimit);
 
+
+export const updateProductInCart = async (userId: string, productId: string, amount: number): Promise<Product> => {
```
Comment (tomporat247) [inline src/cart/cart.ts:38]: all of this is ok, but when you throw gql errors to the user you should use apollo's errors

File: src/graphql/index.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { gql, Config } from "apollo-server";
+import { DocumentNode } from "graphql";
+import { registerNewCart } from "../cart/cart";
+import { productsResolvers } from "./product/product.resolver";
+import { productsSchema } from "./product/product.schema";
+import { ProductTypeDef } from "./product/product.types.schema";
+
+const baseSchema: DocumentNode = gql`
```
Comment (tomporat247) [inline src/graphql/index.ts:8]: I don't think you need this

File: src/web-socket/web-socket.ts [hunk 10]
```diff
@@ -1,40 +1,24 @@
 import { webSocket } from "..";
```
Comment (tomporat247) [inline src/web-socket/web-socket.ts:1]: I don't want want you to waste your time too much on this but you should understand that as of now it seems that if someone will use your http/ws api the gql subscription will never update with the changes

File: src/graphql/index.ts (hunk 9 above)
Comment (alon6699) [reply src/graphql/index.ts:8]: In this project u right but if i want to extend it I have to extend the base query

File: src/graphql/index.ts (hunk 2 above)
Comment (alon6699) [reply src/graphql/index.ts:17]: R u sure is it the best way?

File: src/web-socket/web-socket.ts (hunk 10 above)
Comment (alon6699) [reply src/web-socket/web-socket.ts:1]: Yea 😒 ... du u have a solutuon? I set entire day on this

File: src/cart/cart.ts (hunk 1 above)
Comment (alon6699) [reply src/cart/cart.ts:8]: What do u mean by revert?

**Pull request merged**